alacritty/src/renderer/mod.rs

```diff
@@ -218,3 +219,55 @@ impl Renderer {
         }
     }
 }
+
+struct GlExtensions {
+    extensions: Box<dyn Iterator<Item = Result<&'static str, Utf8Error>>>,
+}
+
+impl GlExtensions {
+    fn new() -> Self {
+        unsafe {
+            let extensions = gl::GetString(gl::EXTENSIONS);
+            if extensions.is_null() {
+                // We're on the core profile, query extension one by one.
+                Self::from_core_profile()
+            } else {
+                Self::from_compatible_profile()
+            }
+        }
+    }
+
+    unsafe fn from_core_profile() -> Self {
+        let mut extensions_number = 0;
+        gl::GetIntegerv(gl::NUM_EXTENSIONS, &mut extensions_number);
+
+        let extensions = (0..extensions_number as gl::types::GLuint).map(|i| {
+            let extension = CStr::from_ptr(gl::GetStringi(gl::EXTENSIONS, i) as *mut _);
+            extension.to_str()
+        });
+
+        GlExtensions { extensions: Box::new(extensions) }
+    }
+
+    unsafe fn from_compatible_profile() -> Self {
+        // SAFETY: OpenGL returns a pointer to a `static` string, so using static lifetime.
+        let extensions = gl::GetString(gl::EXTENSIONS);
+        let extensions: Box<dyn Iterator<Item = Result<&'static str, Utf8Error>>> =
+            match std::mem::transmute::<Result<&'_ str, Utf8Error>, Result<&'static str, Utf8Error>>(
```

Comment:
This looks **very** unsafe. Are you just using `transmute` to extend the lifetime here? Because you should **definitely** not do that.

Reply:
This is safe, since the pointer from C library is pointing to static string by the opengl specification. So the reference will leave to the end of the program.

I'm just using what specification is tell me, if it wasn't safe I wouldn't extended it in the first place?

What would you suggest? To copy everything?

Reply:
Though, I think I don't really need transmute. I'll try without it, but the thing is that all the references are static here anyway.

Reply:
> So the reference will leave to the end of the program.

Is that true for the Rust reference too though? Generally speaking, it seems like a **really** bad idea to extend an existing lifetime with `transmute`. Because even if the compiler doesn't use lifetime information for optimizations, at the very least it would indicate that the original lifetime is incorrect.

Reply:
Well there's no transmute anymore as you can see. The code is marked as `safe`...